File: Site 3/GPT4/imagem_descrita.java
import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.chrome.ChromeDriver;
import org.openqa.selenium.support.ui.ExpectedConditions;
import org.openqa.selenium.support.ui.WebDriverWait;
import org.testng.Assert;
import org.testng.annotations.AfterClass;
import org.testng.annotations.BeforeClass;
import org.testng.annotations.Test;

import java.time.Duration;
import java.util.List;

public class SauceDemoLoginTest {
    private WebDriver driver;
    private WebDriverWait wait;
    private final String baseUrl = "https://www.saucedemo.com/v1/";

    @BeforeClass
    public void setUp() {
        // Set up the ChromeDriver path
        System.setProperty("webdriver.chrome.driver", "path/to/chromedriver");
        driver = new ChromeDriver();
        wait = new WebDriverWait(driver, Duration.ofSeconds(10));
        driver.get(baseUrl);
    }

    @Test
    public void verifyPageStructure() {
        // Verify the presence of username field
        WebElement usernameField = driver.findElement(By.id("user-name"));
        Assert.assertTrue(usernameField.isDisplayed(), "Username field is not displayed");

        // Verify the presence of password field
        WebElement passwordField = driver.findElement(By.id("password"));
        Assert.assertTrue(passwordField.isDisplayed(), "Password field is not displayed");

        // Verify the presence of login button
        WebElement loginButton = driver.findElement(By.id("login-button"));
        Assert.assertTrue(loginButton.isDisplayed(), "Login button is not displayed");

        // Verify the presence of labels
        WebElement acceptedUsernamesLabel = driver.findElement(By.xpath("//div[text()='Accepted usernames are:']"));
        Assert.assertTrue(acceptedUsernamesLabel.isDisplayed(), "Accepted usernames label is not displayed");

        WebElement passwordForAllUsersLabel = driver.findElement(By.xpath("//div[text()='Password for all users:']"));
        Assert.assertTrue(passwordForAllUsersLabel.isDisplayed(), "Password for all users label is not displayed");

        // Verify the presence of four fields under "Accepted usernames are:"
        List<WebElement> acceptedUsernames = driver.findElements(By.xpath("//div[text()='Accepted usernames are:']/following-sibling::div"));
        Assert.assertEquals(acceptedUsernames.size(), 4, "There are not 4 accepted usernames displayed");

        // Verify the presence of one field under "Password for all users:"
        List<WebElement> passwordForAllUsers = driver.findElements(By.xpath("//div[text()='Password for all users:']/following-sibling::div"));
        Assert.assertEquals(passwordForAllUsers.size(), 1, "There is not 1 password for all users displayed");
    }

    @Test
    public void testSuccessfulLogin() {
        // Retrieve valid usernames and password
        List<WebElement> acceptedUsernames = driver.findElements(By.xpath("//div[text()='Accepted usernames are:']/following-sibling::div"));
        String validPassword = driver.findElement(By.xpath("//div[text()='Password for all users:']/following-sibling::div")).getText();

        for (WebElement usernameElement : acceptedUsernames) {
            String validUsername = usernameElement.getText();

            // Perform login
            driver.findElement(By.id("user-name")).clear();
            driver.findElement(By.id("user-name")).sendKeys(validUsername);
            driver.findElement(By.id("password")).clear();
            driver.findElement(By.id("password")).sendKeys(validPassword);
            driver.findElement(By.id("login-button")).click();

            // Verify successful login (assuming redirection to a new page)
            wait.until(ExpectedConditions.urlContains("inventory.html"));
            Assert.assertTrue(driver.getCurrentUrl().contains("inventory.html"), "Login was not successful for username: " + validUsername);

            // Log out to test the next username
            driver.navigate().back();
        }
    }

    @Test
    public void testLoginFailure() {
        // Test with invalid username and valid password
        driver.findElement(By.id("user-name")).clear();
        driver.findElement(By.id("user-name")).sendKeys("invalid_user");
        driver.findElement(By.id("password")).clear();
        driver.findElement(By.id("password")).sendKeys("secret_sauce");
        driver.findElement(By.id("login-button")).click();

        // Verify error message
        WebElement errorMessage = wait.until(ExpectedConditions.visibilityOfElementLocated(By.cssSelector(".error-message-container")));
        Assert.assertTrue(errorMessage.isDisplayed(), "Error message is not displayed for invalid username");

        // Test with valid username and invalid password
        driver.findElement(By.id("user-name")).clear();
        driver.findElement(By.id("user-name")).sendKeys("standard_user");
        driver.findElement(By.id("password")).clear();
        driver.findElement(By.id("password")).sendKeys("wrong_password");
        driver.findElement(By.id("login-button")).click();

        // Verify error message
        errorMessage = wait.until(ExpectedConditions.visibilityOfElementLocated(By.cssSelector(".error-message-container")));
        Assert.assertTrue(errorMessage.isDisplayed(), "Error message is not displayed for invalid password");
    }

    @Test
    public void testErrorMessages() {
        // Test with invalid credentials
        driver.findElement(By.id("user-name")).clear();
        driver.findElement(By.id("user-name")).sendKeys("invalid_user");
        driver.findElement(By.id("password")).clear();
        driver.findElement(By.id("password")).sendKeys("wrong_password");
        driver.findElement(By.id("login-button")).click();

        // Verify error message content
        WebElement errorMessage = wait.until(ExpectedConditions.visibilityOfElementLocated(By.cssSelector(".error-message-container")));
        String expectedErrorMessage = "Epic sadface: Username and password do not match any user in this service";
        Assert.assertEquals(errorMessage.getText(), expectedErrorMessage, "Error message content is incorrect");
    }

    @AfterClass
    public void tearDown() {
        // Close the browser
        if (driver != null) {
            driver.quit();
        }
    }
}

/*
Explanation:
Setup and Teardown: The setUp method initializes the WebDriver and navigates to the login page. The tearDown method closes the browser after tests are completed.
Test Methods:
verifyPageStructure: Checks the presence of all required elements on the page.
testSuccessfulLogin: Iterates over the valid usernames and performs a login with each, verifying successful login by checking the URL.
testLoginFailure: Tests invalid login attempts and verifies that an error message is displayed.
testErrorMessages: Ensures the error message content is correct when invalid credentials are used.
Assertions: Used to validate the presence of elements and the correctness of error messages.
Waits: Explicit waits are used to handle dynamic content loading, ensuring elements are present before interacting with them.
Make sure to replace "path/to/chromedriver" with the actual path to your ChromeDriver executable. Additionally, ensure that the necessary Selenium libraries are included in your project.

*/

File: Códigos/Site 3 - Dynamic/GPT4/imagem_descrita.py
import unittest
from selenium import webdriver
from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC

class DynamicLoadingTests(unittest.TestCase):
    def setUp(self):
        # Initialize the Chrome WebDriver
        self.driver = webdriver.Chrome()
        self.driver.maximize_window()
        # Set base URL
        self.base_url = "https://the-internet.herokuapp.com/dynamic_loading"
        # Set default wait time for explicit waits
        self.wait = WebDriverWait(self.driver, 10)

    def tearDown(self):
        # Close the browser after each test
        self.driver.quit()

    def test_page_structure(self):
        """Test to validate the main page structure and elements"""
        self.driver.get(self.base_url)
        
        # Verify page title
        page_header = self.driver.find_element(By.TAG_NAME, "h3")
        self.assertEqual(page_header.text, "Dynamically Loaded Page Elements")
        
        # Verify presence of both example links
        example_links = self.driver.find_elements(By.TAG_NAME, "a")
        self.assertTrue(any("Example 1: Element on page that is hidden" in link.text for link in example_links))
        self.assertTrue(any("Example 2: Element rendered after the fact" in link.text for link in example_links))

    def test_example1_hidden_element(self):
        """Test Example 1: Element hidden on page"""
        # Navigate to Example 1
        self.driver.get(f"{self.base_url}/1")
        
        # Verify initial state
        start_button = self.wait.until(
            EC.presence_of_element_located((By.CSS_SELECTOR, "#start button"))
        )
        
        # Check if Hello World text is present but hidden
        hidden_element = self.driver.find_element(By.CSS_SELECTOR, "#finish")
        self.assertFalse(hidden_element.is_displayed())
        
        # Click Start button
        start_button.click()
        
        # Wait for loading to complete and verify text
        hello_text = self.wait.until(
            EC.visibility_of_element_located((By.CSS_SELECTOR, "#finish h4"))
        )
        self.assertEqual(hello_text.text, "Hello World!")

    def test_example2_rendered_element(self):
        """Test Example 2: Element rendered after the fact"""
        # Navigate to Example 2
        self.driver.get(f"{self.base_url}/2")
        
        # Verify initial state
        start_button = self.wait.until(
            EC.presence_of_element_located((By.CSS_SELECTOR, "#start button"))
        )
        
        # Verify Hello World text is not present initially
        initial_text_present = self.driver.find_elements(By.CSS_SELECTOR, "#finish h4")
        self.assertEqual(len(initial_text_present), 0)
        
        # Click Start button
        start_button.click()
        
        # Wait for loading to complete and verify text
        hello_text = self.wait.until(
            EC.presence_of_element_located((By.CSS_SELECTOR, "#finish h4"))
        )
        self.assertEqual(hello_text.text, "Hello World!")

if __name__ == "__main__":
    unittest.main()